oml/datasets/vectors_pairs.py

```diff
@@ -0,0 +1,61 @@
+from typing import Dict
+
+import torch
+from torch import Tensor
+from torch.utils.data import DataLoader, Dataset
+from tqdm.auto import tqdm
+
+from oml.const import PAIR_1ST_KEY, PAIR_2ND_KEY
+from oml.interfaces.models import IPairwiseDistanceModel
+
+
+class VectorsPairsDataset(Dataset):
+    """
+    Dataset to iterate over pairs of embeddings.
+
+    """
+
+    def __init__(self, x1: Tensor, x2: Tensor, pair_1st_key: str = PAIR_1ST_KEY, pair_2nd_key: str = PAIR_2ND_KEY):
+        assert x1.shape == x2.shape
+        assert x1.ndim >= 2
+
+        self.pair_1st_key = pair_1st_key
+        self.pair_2nd_key = pair_2nd_key
+
+        self.x1 = x1
+        self.x2 = x2
+
+    def __getitem__(self, idx: int) -> Dict[str, Tensor]:
+        return {self.pair_1st_key: self.x1[idx], self.pair_2nd_key: self.x2[idx]}
+
+    def __len__(self) -> int:
+        return len(self.x1)
+
+
+# fmt: off
+def pairwise_inference(
+        model: IPairwiseDistanceModel,
+        x1: Tensor,
+        x2: Tensor,
+        num_workers: int = 0,
+        batch_size: int = 512,
+        verbose: bool = False
+) -> Tensor:
+    model.eval()
+    dataset = VectorsPairsDataset(x1=x1, x2=x2)
+    loader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, shuffle=False)
+    device = next(model.parameters()).device
+
+    loader = tqdm(loader) if verbose else loader
+
+    outputs = []
+    with torch.no_grad():
+        for batch in loader:
+            x1 = batch[dataset.pair_1st_key].to(device)
+            x2 = batch[dataset.pair_2nd_key].to(device)
+            outputs.append(model(x1=x1, x2=x2))
+
+    return torch.cat(outputs).detach().cpu()
```

Comment:
What about restoring `model.training` state in the end of the function?

Reply:
It looks like it is the only option to guarantee that everything will work as expected. Otherwise someone can just forget to do `model.train()` after the `pairwise_distance` call.

Reply:
Ok. I saw the comment above on it. We have different seeing of the library usage. It looks like you have in mind the config api usage, whereas I have in mind possibility to use parts of this library to build something new, and I wouldn't expect that function could change the `model.training` state and doesn't restore it. I accustomed that a function doesn't change the training flag, and it is my duty to do it, or it changes the flag, but restores it.

Reply:
Okay, I got your point